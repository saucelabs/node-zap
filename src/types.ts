import type { OpenAPIV3 } from 'openapi-types'

export interface ProtocolCommand {
    method: OpenAPIV3.HttpMethods
    endpoint: string
    description: OpenAPIV3.OperationObject
    servers: OpenAPIV3.ServerObject[]
}

export interface Options {
    user: string
    key: string
    region: 'apac'
    headers?: Record<string, string>
}
