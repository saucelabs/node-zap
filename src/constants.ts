import { camelCase } from 'change-case'
import os from 'os'
import type { OpenAPIV3 } from 'openapi-types'

import type { ProtocolCommand, Parameters } from './types'

const { version } = require('../package.json')

export const BINDING_VERSION_NOTE = `node-zap v${version}`

const protocols: OpenAPIV3.Document[] = [
    require('../build/api/zap.json'),
    require('../build/api/sauce.json')
]

export const API_DOMAINS = new Set()
const protocolFlattened: Map<string, ProtocolCommand> = new Map()
for (const { paths, servers, info, components } of protocols) {
    if (!servers) {
        throw new Error(`No "servers" property found in API ${info.title}`)
    }

    for (const [endpoint, methods] of Object.entries(paths as OpenAPIV3.PathsObject<OpenAPIV3.OperationObject>)) {
        for (const [method, description] of Object.entries(methods as Record<OpenAPIV3.HttpMethods, OpenAPIV3.OperationObject>)) {
            if (method === 'parameters') {
                continue
            }

            const commandParams: OpenAPIV3.ParameterObject[] = []
            let params = [
                ...(methods?.parameters || []),
                ...(description.parameters || [])
            ]

            for (const param of params) {
                /**
                 * parameter is referenced, e.g. #/components/parameters/FooBar
                 */
                const refParam = (param as OpenAPIV3.ReferenceObject).$ref
                if (refParam && components?.parameters) {
                    const paramName = refParam.split('/').pop()
                    if (paramName) {
                        commandParams.push(components.parameters[paramName] as OpenAPIV3.ParameterObject)
                    }
                }

                commandParams.push(param as OpenAPIV3.ParameterObject)
            }


            if (!description.operationId) {
                throw new Error(`No "operationId" found in endpoint ${endpoint}`)
            }

            if (!description.tags || !description.tags![0]) {
                throw new Error(`Expected command "${description.operationId}" to be tagged with its domain`)
            }

            API_DOMAINS.add(description.tags![0])
            let commandName = camelCase(description.operationId)
            /**
             * mark commands as depcrecated in the command names
             */
            if (description.deprecated) {
                commandName += 'Deprecated'
            }

            /**
             * ensure we don't double register commands
             */
            /* istanbul ignore if */
            if (protocolFlattened.has(commandName)) {
                throw new Error(`command ${commandName} already registered`)
            }

            const sanitizedParams = commandParams
                .map((param) => {
                    const newParam: Parameters = {
                        nameCamelCased: camelCase(param.name),
                        ...param
                    }
                    return newParam
                })
            protocolFlattened.set(
                commandName,
                {
                    method: method as OpenAPIV3.HttpMethods,
                    endpoint,
                    description,
                    parameters: sanitizedParams,
                    responses: description.responses as OpenAPIV3.ResponsesObject
                }
            )
        }
    }
}

export const PROTOCOL_MAP = protocolFlattened
export const DEFAULT_OPTIONS = {
    user: process.env.SAUCE_USERNAME,
    key: process.env.SAUCE_ACCESS_KEY,
    region: 'apac',
    headers: { 'User-Agent': `zap/v${version} (nodejs ${os.platform()})` },
} as const

export const SYMBOL_INSPECT = Symbol.for('nodejs.util.inspect.custom')
export const SYMBOL_TOSTRING = Symbol.toStringTag
export const SYMBOL_ITERATOR = Symbol.iterator
export const TO_STRING_TAG = 'ZapProxy API Client'

export const USAGE = `ZapProxy Sauce Labs API CLI

Usage: zap <command> [options]`

export const EPILOG = `Copyright ${(new Date()).getUTCFullYear()} Â© Sauce Labs`

export const CLI_PARAMS = [{
    alias: 'h',
    name: 'help',
    description: 'prints help menu'
}, {
    alias: 'u',
    name: 'user',
    description: 'your Sauce Labs username'
}, {
    alias: 'k',
    name: 'key',
    description: 'your Sauce Labs user key'
}, {
    alias: 'r',
    name: 'region',
    default: DEFAULT_OPTIONS.region,
    description: 'your Sauce Labs datacenter region, the following regions are available: `us-west-1` (short `us`), `eu-central-1` (short `eu`)'
}]

export const SESSION_SUFFIXES = [
    '.session',
    '.session.data',
    '.session.lck',
    '.session.log',
    '.session.properties',
    '.session.script'
]

export const REPORT_COMMANDS = [
    'xmlreport',
    'htmlreport',
    'mdreport'
]
