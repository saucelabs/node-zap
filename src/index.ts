import got from 'got'
import tunnel from 'tunnel'
import { camelCase } from 'change-case'
import type { OpenAPIV3 } from 'openapi-types'

import { toString, getRegionSubDomain } from './utils'
import {
    PROTOCOL_MAP, DEFAULT_OPTIONS, SYMBOL_INSPECT, SYMBOL_TOSTRING,
    SYMBOL_ITERATOR, TO_STRING_TAG, API_DOMAINS
} from './constants'
import { Options, ProtocolCommand } from './types'

export default class Zap {
    public user: string
    public region: string
    public sessionId?: string

    private _options: Options
    private _accessKey: string
    private _api: typeof got
    private _proxy: Zap

    constructor (public options?: Partial<Options>) {
        const opts = Object.assign({}, DEFAULT_OPTIONS, options || {})

        if (!opts.user || !opts.key) {
            throw new Error('"user" and "key" parameters')
        }

        this._options = opts as Options
        this.region = opts.region
        this.user = opts.user

        this._accessKey = this._options.key
        this._api = got.extend({
            headers: opts.headers,
            username: opts.user,
            password: opts.key,
            agent: {
                http: tunnel.httpsOverHttps({
                    proxy: {
                        host: `zap.${getRegionSubDomain(this._options)}.saucelabs.com`,
                        port: 443
                    }
                })
            }
        })

        this._proxy = new Proxy({
            user: this.user,
            key: `XXXXXXXX-XXXX-XXXX-XXXX-XXXXXX${(this._accessKey || '').slice(-6)}`,
            region: this._options.region,
            headers: this._options.headers
        } as any, { get: this._get.bind(this) })

        // @ts-ignore
        return this._proxy
    }

    private _get (scope: { domain: string }, propName: string | symbol): any {
        /**
         * print to string output
         * https://nodejs.org/api/util.html#util_util_inspect_custom
         */
        /* istanbul ignore next */
        if (propName === SYMBOL_INSPECT || propName === 'inspect') {
            return () => toString(this)
        }

        /**
         * print to string tag
         * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag
         */
        if (propName === SYMBOL_TOSTRING) {
            return TO_STRING_TAG
        }

        /**
         * return instance iterator
         * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator
         */
        if (propName === SYMBOL_ITERATOR || typeof propName !== 'string') {
            return
        }

        /**
         * allow to return publicly registered class properties
         */
        if ((this as any)[propName]) {
            return !propName.startsWith('_') ? (this as any)[propName] : undefined
        }

        /**
         * check if domain is being accessed
         */
        if (API_DOMAINS.has(propName)) {
            return new Proxy({ domain: propName }, { get: this._get.bind(this) })
        }

        const commandNameAsView = `${scope.domain}View${propName.slice(0, 1).toUpperCase()}${propName.slice(1)}`
        const commandNameAsAction = `${scope.domain}Action${propName.slice(0, 1).toUpperCase()}${propName.slice(1)}`
        const commandName = PROTOCOL_MAP.has(commandNameAsView)
            ? commandNameAsView
            : PROTOCOL_MAP.has(commandNameAsAction)
                ? commandNameAsAction
                : scope.domain === 'session' ? propName : null
        if (!commandName) {
            /**
             * just return if propName is a symbol (Node 8 and lower)
             */
            /* istanbul ignore next */
            if (typeof propName !== 'string') {
                return
            }
            throw new Error(`Couldn't find API endpoint for command "${propName}"`)
        }

        /**
         * ensure user is authenticated
         */
        // if (scope.domain !== 'session' && !this.sessionId) {
        //     throw new Error(`Couldn't call command "${propName}", reason: not authenticated! Please call \`zap.session.new({ ... })\` first to authenticate with Sauce Labs cloud.`)
        // }

        return async (args: Record<string, any>) => {
            const { method, endpoint, parameters } = PROTOCOL_MAP.get(commandName) as ProtocolCommand

            /**
             * validate parameters
             */
            const bodyMap = new Map()
            for (const param of parameters) {
                const userParam = args[param.nameCamelCased]
                const type = (param.schema as OpenAPIV3.SchemaObject).type!.replace('integer', 'number')

                /**
                 * skip param if not set
                 */
                if (typeof userParam === 'undefined') {
                    if (param.required) {
                        throw new Error(`Missing required parameter "${param.name}" for command "${commandName}"`)
                    }

                    continue
                }

                if (typeof userParam !== type) {
                    throw new Error(`Expected parameter for param '${param.name}' from type '${type}', found '${typeof userParam}'`)
                }

                bodyMap.set(param.name, userParam)
            }

            /**
             * get request body by using the body parameter or convert the parameter
             * map into json object
             */
            const body = [...bodyMap.entries()].reduce((e: any, [k, v]) => {
                e[k] = v
                return e
            }, {})

            /**
             * make request
             */
            const uri = `https://zap.${getRegionSubDomain(this._options)}.saucelabs.com${endpoint}`
            try {
                const response = await this._api[method as 'get'](uri, {
                    ...(
                        method === 'get'
                            ? { searchParams: body }
                            : { json: body }
                    ),
                    responseType: 'json',
                    ...(!this.sessionId ? {} : {
                        headers: { 'x-zap-api-key': this.sessionId }
                    })
                }) as any

                /**
                 * attach session id to the scope
                 */
                if (typeof response.body.sessionId) {
                    this.sessionId = response.body.sessionId
                }

                return response.body
            } catch (err) {
                throw new Error(`Failed calling ${propName as string}: ${err.message}, ${err.response && err.response.body}`)
            }
        }
    }
}
