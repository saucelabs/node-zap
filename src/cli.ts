import yargs from 'yargs'

import { USAGE, CLI_PARAMS, EPILOG, PROTOCOL_MAP, DEFAULT_OPTIONS, BINDING_VERSION_NOTE } from './constants'
import SauceLabs from './index'
import { OpenAPIV3 } from 'openapi-types'
import { camelCase } from 'change-case'

export const run = () => {
    let argv = yargs.usage(USAGE)
        .epilog(EPILOG)
        .demandCommand()
        .commandDir('commands')
        .help()
        .version(BINDING_VERSION_NOTE)

    for (const [commandName, options] of PROTOCOL_MAP) {
        const command = `${commandName} ${options.parameters.map((p) => (
            p.required ? `<${p.name}>` : `[${p.name}]`
        )).join(' ')}`

        const description = (
            options.description.summary ||
            options.description.description ||
            'Unknown description'
        )
        yargs.command(command.trim(), description, (yargs) => {
            for (const param of options.parameters) {
                const schema = param.schema as OpenAPIV3.SchemaObject
                const paramDescription = {
                    describe: param.description,
                    type: schema.type
                } as any

                if (typeof schema.default !== 'undefined') {
                    paramDescription.default = schema.default
                }

                yargs.positional(param.name, paramDescription)
            }
        }, async (argv) => {
            const { user, key, region } = Object.assign({}, DEFAULT_OPTIONS, argv)

            if (!user || !key) {
                console.error('"user" and "key" parameters required')
                return process.env.JEST_WORKER_ID
                    ? null
                    : process.exit(1)
            }

            const api = new SauceLabs({ user, key, region }) as any
            const command = PROTOCOL_MAP.get(commandName)
            // eslint-disable-next-line no-unused-vars
            const [domain, shortCommandName] = commandName.split(/(Action|View)/)

            if (!command) {
                throw new Error(`Command "${commandName}" not found`)
            }

            try {
                const result = await api[domain][camelCase(shortCommandName)](argv)

                if (typeof result === 'object') {
                    // eslint-disable-next-line no-console
                    return console.log(JSON.stringify(result, null, 4))
                }

                // eslint-disable-next-line no-console
                console.log(result)
            } catch (error) {
                // eslint-disable-next-line no-console
                console.error(error)
                /* istanbul ignore next */
                return process.env.JEST_WORKER_ID
                    ? error
                    : process.exit(1)
            }

            /**
             * only return for testing purposes
             */
            /* istanbul ignore next */
            if (process.env.JEST_WORKER_ID) {
                return api
            }
        })
    }

    /**
     * populate cli arguments
     */
    for (const param of CLI_PARAMS) {
        argv = argv.option(param.name, param)
    }

    return argv.argv
}
