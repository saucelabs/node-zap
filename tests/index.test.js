import util from 'util'
import got from 'got'
import { spawn } from 'child_process'
import FormData from 'form-data'

import SauceLabs from '../src'
import { instances } from 'bin-wrapper'
import versions from './__responses__/versions.json'
import {DEFAULT_SAUCE_CONNECT_VERSION} from '../src/constants'

jest.mock('fs')
const fs = require('fs')
fs.promises = {
    stat: jest.fn().mockReturnValue(Promise.resolve({ size: 123 }))
}

jest.mock('child_process', () => {
    const EventEmitter = require('events')
    const stdoutEmitter = new EventEmitter()
    const stderrEmitter = new EventEmitter()
    const spawnMock = {
        pid: 123,
        stderr: stderrEmitter,
        stdout: stdoutEmitter,
        on: jest.fn()
    }
    const spawn = jest.fn().mockReturnValue(spawnMock)
    return { spawn }
})

const stdoutEmitter = spawn().stdout
const stderrEmitter = spawn().stderr
const origKill = ::process.kill
beforeEach(() => {
    spawn.mockClear()
    process.kill = jest.fn()
    // clean instances array
    instances.splice(0,instances.length)
})
test('should be inspectable', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    /**
     * we can't use snapshotting here as the result varies
     * between different node versions
     */
    expect(util.inspect(api)).toContain(`{
  username: 'foo',
  key: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXbar',
  region: 'us',
  headless: false,
  proxy: undefined`)
})

test('should expose a webdriverEndpoint', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(api.webdriverEndpoint)
        .toBe('https://ondemand.us-west-1.saucelabs.com/')

    const api2 = new SauceLabs({ user: 'foo', key: 'bar', region: 'eu' })
    expect(api2.webdriverEndpoint)
        .toBe('https://ondemand.eu-central-1.saucelabs.com/')

    const api3 = new SauceLabs({ user: 'foo', key: 'bar', region: 'eu', headless: true })
    expect(api3.webdriverEndpoint)
        .toBe('https://ondemand.us-east-1.saucelabs.com/')

    const api4 = new SauceLabs({ user: 'foo', key: 'bar', region: 'us-central-3' })
    expect(api4.webdriverEndpoint)
        .toBe('https://ondemand.us-central-3.saucelabs.com/')
})

test('should have to string tag', () => {
    expect(Object.prototype.toString.call(
        new SauceLabs({ user: 'foo', key: 'bar' })
    )).toBe('[object SauceLabs API Client]')
})

test('should not provide an iterator', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(() => [...api]).toThrow('is not iterable')
})

test('should return public properties', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(api.region).toBe('us')
    expect(api.username).toBe('foo')
    expect(api._accessKey).toBe(undefined)
})

test('should return nothing if Symbol was accessed', () => {
    const sym = Symbol('foo')
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(typeof api[sym]).toBe('undefined')
})

test('should grab username and access key from env variable', () => {
    jest.resetModules()
    process.env.SAUCE_USERNAME = 'barfoo'
    process.env.SAUCE_ACCESS_KEY = 'foobar'
    const SauceLabsNew = require('../src').default
    const api = new SauceLabsNew()
    expect(util.inspect(api))
        .toContain('username: \'barfoo\'')
    expect(util.inspect(api))
        .toContain('key: \'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXfoobar\'')
})

test('should grab http proxy from env variable', () => {
    jest.resetModules()
    process.env.HTTP_PROXY = 'http://my.proxy.com:8080'
    const SauceLabsNew = require('../src').default
    const api = new SauceLabsNew()
    expect(util.inspect(api))
        .toMatch(/proxy: '??http:\/\/my\.proxy\.com:8080/)
    delete process.env.HTTP_PROXY
})

test('should grab https proxy from env variable', () => {
    jest.resetModules()
    process.env.HTTPS_PROXY = 'https://my.proxy.com:443'
    const SauceLabsNew = require('../src').default
    const api = new SauceLabsNew()
    expect(util.inspect(api))
        .toMatch(/proxy: '??https:\/\/my\.proxy\.com:443/)
})

test('should throw if API command is unknown', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(() => api.doSomethingCool(123, { foo: 'bar' }))
        .toThrow('Couldn\'t find API endpoint for command "doSomethingCool"')
})

test('should allow to call an API method with param in url', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.getUserConcurrency('someuser')
    expect(got.mock.calls[0][0])
        .toBe('https://api.us-west-1.saucelabs.com/rest/v1.2/users/someuser/concurrency')
})

test('should allow to call an API method with param as option', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.listJobs('someuser', {
        limit: 123,
        full: true
    })

    const uri = got.mock.calls[0][0]
    const req = got.mock.calls[0][1]
    expect(uri).toBe('https://api.us-west-1.saucelabs.com/rest/v1.1/someuser/jobs')
    expect(req.searchParams).toEqual({
        limit: 123,
        full: true
    })
})

test('should allow to make a request with body param', async () =>  {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.updateJob('foobaruser', '690c5877710c422d8be4c622b40c747f', { passed: true })

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: true })
})

test('should allow to make a request with body param via CLI call', async () =>  {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.updateJob('foobaruser', '690c5877710c422d8be4c622b40c747f', '{ "passed": false }')

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: false })
})

test('should update RDC job status with body param', async () =>  {
    const api = new SauceLabs({})
    await api.updateTest('89ec3cca-7092-41f1-8037-d035579fb8d1', { passed: true })

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: true })
})

test('should update RDC job status with body param via CLI call', async () =>  {
    const api = new SauceLabs({})
    await api.updateTest('89ec3cca-7092-41f1-8037-d035579fb8d1', '{ "passed": false }')

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: false })
})

test('should fail if param has wrong type', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs(123, {
        limit: 123,
        full: true
    }).catch((err) => err)
    expect(error.message).toBe('Expected parameter for url param \'username\' from type \'string\', found \'number\'')
})

test('should fail if option has wrong type', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs('someuser', {
        limit: '123',
        full: true
    }).catch((err) => err)
    expect(error.message).toBe('Expected parameter for option \'limit\' from type \'number\', found \'string\'')
})

test('should handle error case', async () => {
    const response = new Error('Not Found')
    response.statusCode = 404
    response.body = { message: 'Not Found' }
    got.get.mockReturnValueOnce(Promise.reject(response))
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs('someuser', {
        limit: 123,
        full: true
    }).catch((err) => err)
    expect(error.message).toContain('Failed calling listJobs: Not Found')
})

test('should be able to download assets', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json')
    const uri = got.get.mock.calls[0][0]
    expect(uri).toContain('https://assets.saucelabs.com/jobs/some-id/performance.json')
    expect(uri).toContain('auth=a2600100e3d1990721be97c093f64567')
})

test('should handle errors when downloading assets', async () => {
    const response = new Error('Not Found')
    response.statusCode = 404
    response.body = { message: 'Not Found' }
    got.get.mockReturnValueOnce(Promise.reject(response))

    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.downloadJobAsset('some-id', 'performance.json').catch((err) => err)
    expect(error.message).toBe('There was an error downloading asset performance.json: Not Found')
})

test('should parse text responses if headers expect json', async () => {
    const reqRespond = { foo: 'bar' }
    got.get.mockReturnValueOnce(Promise.resolve({
        headers: {
            'content-type': 'application/json'
        },
        body: JSON.stringify(reqRespond)
    }))
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const result = await api.downloadJobAsset('some-id', 'performance.json')
    expect(result).toEqual(reqRespond)
})

test('should fail if parameters are not given properly', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = new Error('You need to define a job id and the file name of the asset as a string')
    await expect(api.downloadJobAsset()).rejects.toEqual(error)
    await expect(api.downloadJobAsset('foo')).rejects.toEqual(error)
    await expect(api.downloadJobAsset(123, 'bar')).rejects.toEqual(error)
})

test('should support proxy options', async () => {
    const proxy = 'http://my.proxy.com:8080'
    const api = new SauceLabs({ user: 'foo', key: 'bar', proxy })
    await api.downloadJobAsset('some-id', 'performance.json')
    const requestOptions = got.extend.mock.calls[1][0]

    await expect(requestOptions.agent).toBeDefined()
})

test('should put asset into file as binary', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json', { filepath: '/asset.json' })
    expect(fs.writeFileSync).toBeCalledWith('/asset.json', undefined, { encoding: 'binary' })
})

test('should put asset into file as json file', async () => {
    got.setHeader({ 'content-type': 'application/json' })
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json', { filepath: '/asset.json' })
    expect(fs.writeFileSync).toBeCalledWith('/asset.json', undefined, { encoding: 'utf8' })
})

test('should allow to upload files', async () => {
    fs.createReadStream.mockReturnValue({
        name: '/somefile',
        path: 'somepath',
    })

    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const body = { foo: 'bar' }
    got.mockReturnValue(Promise.resolve({
        headers: {
            'content-type': 'application/json'
        },
        body: JSON.stringify(body)
    }))
    const result = await api.uploadJobAssets('some-id', {
        files: [
            'log.json',
            'selenium-server.json',
            {
                filename: 'foobar.json',
                data: { foo: 'bar' }
            }
        ]
    })

    const { instances } = new FormData()
    expect(instances[0].append).toBeCalledTimes(3)
    expect(instances[0].append).toBeCalledWith(
        'file[]',
        { name: '/somefile', path: 'somepath' },
        {
            contentType: 'text/plain',
            filename: 'log.json',
            filepath: expect.any(String),
            knownLength: 123
        }
    )
    expect(instances[0].append).toBeCalledWith('file[]', Buffer.from(JSON.stringify({ foo: 'bar' })), 'foobar.json')

    const uri = got.mock.calls[0][0]
    expect(uri).toBe('https://api.us-west-1.saucelabs.com/v1/testcomposer/jobs/some-id/assets')

    expect(result).toEqual(body)
})

test('should throw if custom error if upload fails', async () => {
    fs.createReadStream.mockReturnValue({
        name: '/somefile',
        path: 'somepath',
    })

    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    got.mockReturnValue(Promise.reject(new Error('uups')))
    const result = await api.uploadJobAssets('some-id', {
        files: ['log.json', '/selenium-server.json']
    })
        .catch((err) => err)

    expect(result.message).toBe('There was an error uploading assets: uups')
})

test('should not even try to upload if no files were selected', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const result = await api.uploadJobAssets('some-id')
        .catch((err) => err)

    expect(result.message).toBe('No files to upload selected')
})

test('should fail if file parameter is invalid', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const result = await api.uploadJobAssets('some-id', {
        files: [{ foo: 'bar' }]
    }).catch((err) => err)

    expect(result.message).toContain('Invalid file parameter')
})

test('should contain custom headers', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar', headers: { 'user-agent': 'foo' } })
    got.mockReturnValue(Promise.resolve({
        body: JSON.stringify({ foo: 'bar' })
    }))
    await api.updateTest('123', '{ "passed": false }')
    const requestOptions = got.extend.mock.calls[0][0]
    expect(requestOptions.headers).toMatchSnapshot()
})

describe('startSauceConnect', () => {
    it('should start sauce connect with proper parsed args', async () => {
        const logs = []
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        setTimeout(() => stdoutEmitter.emit('data', 'Sauce Connect is up, you may start your tests'), 50)
        await api.startSauceConnect({
            scVersion: '1.2.3',
            tunnelIdentifier: 'my-tunnel',
            'proxy-tunnel': 'abc',
            logger: (log) => logs.push(log)
        })
        expect(spawn).toBeCalledTimes(1)
        expect(spawn.mock.calls).toMatchSnapshot()

        expect(logs).toHaveLength(1)
        expect(instances).toHaveLength(1)
        expect(instances[0].dest.mock.calls[0][0].endsWith('.sc-v1.2.3'))
            .toBe(true)
    })

    it('should start sauce connect with latest version if no version is specified in the args', async () => {
        const logs = []
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        got.mockReturnValue(Promise.resolve({
            body: {
                data: {
                    ...versions
                }
            }
        }))
        setTimeout(() => stdoutEmitter.emit('data', 'Sauce Connect is up, you may start your tests'), 50)
        await api.startSauceConnect({
            tunnelIdentifier: 'my-tunnel',
            'proxy-tunnel': 'abc',
            logger: (log) => logs.push(log)
        })
        expect(instances[0].dest.mock.calls[0][0].endsWith('.sc-v1.2.4'))
            .toBe(true)
    })

    it('should start sauce connect with fallback default version in case the call to the API failed', async () => {
        const logs = []
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        got.mockImplementation(() => { throw new Error('Endpoint not available!') })
        setTimeout(() => stdoutEmitter.emit('data', 'Sauce Connect is up, you may start your tests'), 50)
        await api.startSauceConnect({
            tunnelIdentifier: 'my-tunnel',
            'proxy-tunnel': 'abc',
            logger: (log) => logs.push(log)
        })
        expect(instances[0].dest.mock.calls[0][0].endsWith(`.sc-v${DEFAULT_SAUCE_CONNECT_VERSION}`))
            .toBe(true)
    })

    it('should properly fail if connection could not be established', async () => {
        const errMessage = 'Sauce Connect could not establish a connection'
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        setTimeout(() => stdoutEmitter.emit('data', errMessage), 50)
        const err = await api.startSauceConnect({
            scVersion: '1.2.3',
            tunnelIdentifier: 'my-tunnel',
            'proxy-tunnel': 'abc'
        }).catch((err) => err)
        expect(err.message).toBe(errMessage)
    })

    it('should close sauce connect', async () => {
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        setTimeout(() => stdoutEmitter.emit('data', 'Sauce Connect is up, you may start your tests'), 50)
        const sc = await api.startSauceConnect({ tunnelIdentifier: 'my-tunnel' }, true)
        setTimeout(() => {
            sc.cp.stdout.emit('data', 'Some other message')
            sc.cp.stdout.emit('data', 'Goodbye')
        }, 100)
        await sc.close()
        expect(process.kill).toBeCalledWith(123, 'SIGINT')
    })

    it('should fail if stderr is emitted', async () => {
        const api = new SauceLabs({ user: 'foo', key: 'bar', headless: true })
        setTimeout(() => stderrEmitter.emit('data', 'Uuups'), 50)
        const res = await api.startSauceConnect({ tunnelIdentifier: 'my-tunnel' }).catch((err) => err)
        expect(res).toEqual(new Error('Uuups'))
    })

    it('should not overwrite rest-url if given as a parameter', async () => {
        const api = new SauceLabs({ user: 'foo', key: 'bar'})
        setTimeout(() => stdoutEmitter.emit('data', 'Sauce Connect is up, you may start your tests'), 50)
        await api.startSauceConnect({
            tunnelIdentifier: 'my-tunnel',
            restUrl: 'https://us1.api.testobject.com/sc/rest/v1'
        })
        expect(spawn.mock.calls).toMatchSnapshot()
    })
})

test('should output failure msg for createJob API', async () => {
    const response = new Error('Response code 422 (Unprocessable Entity)')
    response.statusCode = 422
    response.response = {body: 'empty framework'}
    got.post.mockReturnValue(Promise.reject(response))

    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.createJob({framework: ''}).catch((err) => err)

    expect(error.message).toBe('Failed calling createJob: Response code 422 (Unprocessable Entity), empty framework')
})

afterEach(() => {
    fs.writeFileSync.mockClear()
    got.mockClear()
    got.extend.mockClear()
    got.put.mockClear()
    got.get.mockClear()
    process.kill = origKill
})
