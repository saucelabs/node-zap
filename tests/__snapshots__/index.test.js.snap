// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should contain custom headers 1`] = `
Object {
  "Authorization": "Basic Zm9vOmJhcg==",
  "user-agent": "foo",
}
`;

exports[`startSauceConnect should not overwrite rest-url if given as a parameter 1`] = `
Array [
  Array [
    "/foo/bar",
    Array [
      "--tunnel-identifier=my-tunnel",
      "--rest-url=https://us1.api.testobject.com/sc/rest/v1",
      "--user=foo",
      "--api-key=bar",
    ],
  ],
]
`;

exports[`startSauceConnect should start sauce connect with proper parsed args 1`] = `
Array [
  Array [
    "/foo/bar",
    Array [
      "--proxy-tunnel=abc",
      "--tunnel-identifier=my-tunnel",
      "--user=foo",
      "--api-key=bar",
      "--rest-url=https://us-east-1.saucelabs.com/rest/v1",
    ],
  ],
]
`;
